Implement user authentication and signup

aonomike opened the pull request:
#### Fixes issue #issue <!-- Add Issue Number here -->
<!-- Describe What you did in this pr -->
- Use devise-jwt gem
- Implement signup
- Implement Signin
- Implement signout

- [ ] Tests all green tests and all builds passes :green_heart:
- [x] PR well-named, issue linked :paperclip: and no merge conflicts.
- [ ] Request review from other contributors

##### Screenshots(if appropriate)

aonomike committed:
Start implement user authentication

- Use devise-jwt gem
- Still getting some errors
- Run rubocop on a number of the files

aonomike committed:
Complete devise-jwt setup

- Complete setup of devise jwt
- Add dotenv-rails gem to help with reading the key

aonomike committed:
Merge branch 'develop' of github.com:AgileVentures/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

aonomike committed:
Fix jwt token typo

aonomike committed:
Start implementing tests for /login

- Test that it reurns 500
- Test still fails

aonomike committed:
Correctly set up devise JWT

- Test that request returns 200 passes

aonomike committed:
Add tests for Login and Logout

- Test jwt token authentication

aonomike committed:
Fix failing event specs

- Add authentication headers to the event specs
- For posts, we removed Content-Type from the authentication headers due to errors that were caused by reasons given in [this stackoverflow link](https://stackoverflow.com/questions/49053954/testing-post-actiondispatchhttpparametersparseerror-765)

Co-authored-by: Gideon Kimutai <gideon.kimutai.kim@gmail.com>
Co-authored-by: Herman Lule <luleherll@gmail.com>

aonomike committed:
Add dotenv rails gem

aonomike committed:
Merge branch 'develop' into 25_signup_endpoint-devise-jwt

aonomike committed:
Implement devise jwt signup

aonomike committed:
Add RegistrationController tests

aonomike committed:
Add tests for registration controller

- Test the create method for registration controller

aonomike committed:
Merge branch 'develop' of github.com:AgileVentures/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

aonomike committed:
Merge branch '25_signup_endpoint-devise-jwt' of github.com:aonomike/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

aonomike committed:
Test that signup returns a token on success

aonomike committed:
Refactor to seperate error tests in a describe

Co-authored-by: Gideon Kimutai <gideon.kimutai.kim@gmail.com>
Co-authored-by: StlMaris123 <stellamarisnjage@gmail.com>

aonomike committed:
Test that registration returns a valid jwt token

gr1d99 commented on app/controllers/events_controller.rb:
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class EventsController < ApplicationController
+  before_action :authenticate_user!
```
add new line here

gr1d99 commented on app/models/jwt_blacklist.rb:
```diff
@@ -0,0 +1,7 @@
+class JwtBlacklist < ApplicationRecord
+  include Devise::JWT::RevocationStrategies::Blacklist
+
+  self.table_name = 'jwt_blacklist'
```
this should be in line 4

gr1d99 commented on config/application.rb:
```diff
@@ -33,5 +35,15 @@ class Application < Rails::Application
     config.api_only = true
+    config.middleware.insert_before 0, Rack::Cors do
+      allow do
+        origins 'http://your.frontend.domain.com'
```
origins value does no seem right, ideally, this should be defined for specific env ie local dev, staging and production, I would strongly suggest we move this to env specific files

gr1d99 commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,310 @@
+  # config.sign_in_after_change_password = true
+
+  config.jwt do |jwt|
+    jwt.secret = ENV['DEVISE_JWT_SECRET_KEY']
```
I suggest we explore rails encrypted credentials

gr1d99 commented on spec/requests/events/events_spec.rb:
```diff
@@ -3,13 +3,20 @@
 RSpec.describe EventsController do
+  let(:user) { create(:user) }
+  let(:headers) { { 'Accept' => 'application/json', 'Content-Type' => 'application/json' } }
+  let(:auth_headers) { Devise::JWT::TestHelpers.auth_headers(headers, user) }
```
We can create a helper method for flexibility

gr1d99 commented on spec/requests/events/events_spec.rb:
```diff
@@ -3,13 +3,20 @@
   let(:endpoint) { '/events' }
   let(:event) { create(:event) }
   let(:invalid_event) { create(:invalid_event) }
+  let(:custom_headers) do
```
the same applies here

aonomike replied:
Which file would you suggest we put the helper? I was thinking to avoid users moving to a different file to looks for methods which can be implemented as one liners?

gr1d99 commented on spec/requests/login/authentication_spec.rb:
```diff
@@ -0,0 +1,53 @@
+  context 'when login params are incorrect' do
+    before { post url }
+
+    it 'returns unathorized status' do
```
This can be refactored [https://www.youtube.com/watch?v=f0I6a0AntP0](https://www.youtube.com/watch?v=f0I6a0AntP0)

aonomike replied:
create a refactoring ticket for this.

aonomike committed:
Update to resolve review comments

aonomike committed:
Resolve conflict

aonomike committed:
Resolve conflict

wutt77 reviewed (approved):
LGTM

aonomike committed:
Merge branch 'develop' of github.com:AgileVentures/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

aonomike committed:
Merge branch 'develop' into 25_signup_endpoint-devise-jwt

mattwr18 committed:
Update Gemfile to version dry-configurable

- was causing NoMethodError: undefined method `map' for nil:NilClass

/home/travis/.rvm/gems/ruby-2.6.2/gems/warden-jwt_auth-0.3.6/lib/warden/jwt_auth.rb:89:in `upcase_first_items'

- found a solution here https://stackoverflow.com/questions/60470873/why-i-have-an-error-when-i-generate-jwt-blacklist-model-for-rails-5/60491838#60491838?newreg=7071f31f48194c26875f49f9b463c008

Co-authored-by: Mike Aono <aonomike@gmail.com>

aonomike committed:
Merge branch '25_signup_endpoint-devise-jwt' of github.com:aonomike/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

mattwr18 committed:
Run rubocop -a

mattwr18 committed:
Get working with WebsiteOne-FE locally

- we can discuss the routes, and config in application.rb, but this is a
  start, and one can successfully sign up

aonomike committed:
Resolve conflicts

aonomike committed:
Merge branch '25_signup_endpoint-devise-jwt' of github.com:aonomike/WebsiteOneBackendApi into 25_signup_endpoint-devise-jwt

aonomike committed:
Fix failing tests

Co-authored-by: Riccardo Giomi <contact@riccardogiomi.com>
Co-authored-by: nicoale_ <nicoale8786@gmail.com>
Co-authored-by: Edwin <fadhilied00@gmail.com>
Co-authored-by: Gideon Kimutai <gideon.kimutai.kim@gmail.com>
